### PR TITLE
UI render

### PR DESCRIPTION
**Pull request by mkrump**
https://trello.com/c/mjXkBURK/26-3-75-ui-render-components-will-be-parameterized-by-loaded-ui-choice-4-5

### CODE REVIEW

**Comment by andarcabrera** [comment]
You have a few commits named `minor`. Please squash commits WIP (work in progress) commits when you make the PR for a clean commit history. 

Also 👍 on the detailed commit historyy

File: Tests/MenuTests.cs [hunk 1]
```diff
@@ -12,22 +12,22 @@ public class RunMenus
...
-                Menu menu = new Menu(new List<ISubMenu> {mock.Object, mock.Object});
+                Menu menu = new Menu(new List<IPrompt> {mock.Object, mock.Object});
                 Assert.Equal(new List<string> {"ARG", "ARG"}, menu.RunSubMenus());
             }
 
             [Fact]
             public void SubMenuTitleMenu()
             {
-                var mock = new Mock<ISubMenu>();
+                var mock = new Mock<IPrompt>();
```
Comment (andarcabrera) [inline Tests/MenuTests.cs]: `mock`is a very general name. it's easy to follow here as it's a small test, but in general go with more descriptive names (no need to change, just FYI)

File: TicTacToe/MenuConfigurations.cs [hunk 2]
```diff
@@ -3,38 +3,51 @@
...
+                string uiChoiceMessage = string.Format("Choose a UI ({0})", configFilesCSV);
+                string uiChoiceErrorMessage = string.Format("Invalid Input. Expected one of the following: {0}",
+                    configFilesCSV);
+                Regex validInput = new Regex(string.Format(@"{0}", string.Join("|", configFiles)));
+                var validFileNames = new RegexInputRule(validInput, uiChoiceErrorMessage);
+                JsonSerializer serializer = new JsonSerializer {MissingMemberHandling = MissingMemberHandling.Error};
+                UIConfigLoader uiConfigLoader = new UIConfigLoader(serializer);
+                var validConfig = new UIConfigIsValid(uiConfigLoader);
+                List<IValidation> setupMenuValidtionRules = new List<IValidation> {validFileNames, validConfig};
+                Prompt gridSizeMenu = new Prompt(
```
Comment (andarcabrera) [inline TicTacToe/MenuConfigurations.cs]: is this `gridSizeMenu` or `UIMenu`?

File: TicTacToe/MenuConfigurations.cs [hunk 3]
```diff
@@ -3,38 +3,51 @@
...
+                var configFilesCSV = string.Join(", ", configFiles);
+                string uiChoiceMessage = string.Format("Choose a UI ({0})", configFilesCSV);
+                string uiChoiceErrorMessage = string.Format("Invalid Input. Expected one of the following: {0}",
+                    configFilesCSV);
+                Regex validInput = new Regex(string.Format(@"{0}", string.Join("|", configFiles)));
+                var validFileNames = new RegexInputRule(validInput, uiChoiceErrorMessage);
+                JsonSerializer serializer = new JsonSerializer {MissingMemberHandling = MissingMemberHandling.Error};
+                UIConfigLoader uiConfigLoader = new UIConfigLoader(serializer);
+                var validConfig = new UIConfigIsValid(uiConfigLoader);
+                List<IValidation> setupMenuValidtionRules = new List<IValidation> {validFileNames, validConfig};
```
Comment (andarcabrera) [inline TicTacToe/MenuConfigurations.cs]: typo `setupMenuValidtionRules`

File: TicTacToe/MenuConfigurations.cs [hunk 4]
```diff
@@ -3,38 +3,51 @@
...
     {
         public class Menus
         {
+            public static List<string> FilesToMessage(string path)
+            {
+                string[] files = Directory.GetFiles(path);
+                return files.Select(Path.GetFileNameWithoutExtension).ToList();
+            }
+
             public static Menu SetupUIMenu(TextReader textReader, TextWriter textWriter)
```
Comment (andarcabrera) [inline TicTacToe/MenuConfigurations.cs]: This method is kind of big. Can you extract some logic in other methods?

File: TicTacToe/UIObejcts.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
+﻿using System;
```
Comment (andarcabrera) [inline TicTacToe/UIObejcts.cs]: File name typo `UIObejcts.cs`

File: TicTacToe/MenuConfigurations.cs (hunk 2 above)
Comment (mkrump) [inline TicTacToe/MenuConfigurations.cs]: Should be UI menu

File: TicTacToe/MenuConfigurations.cs (hunk 3 above)
Comment (mkrump) [inline TicTacToe/MenuConfigurations.cs]: Fixed.

File: TicTacToe/MenuConfigurations.cs (hunk 4 above)
Comment (mkrump) [inline TicTacToe/MenuConfigurations.cs]: I ended up putting all of these files into their own directory, with one file per menu. Hopefully that is a little better.

File: TicTacToe/UIObejcts.cs (hunk 5 above)
Comment (mkrump) [inline TicTacToe/UIObejcts.cs]: Fixed.

**Pull request closed without merging**